src/coreclr/jit/importer.cpp

```diff
@@ -8083,133 +8083,56 @@ GenTree* Compiler::impInitClass(CORINFO_RESOLVED_TOKEN* pResolvedToken)
-                    *(unsigned __int16*)p_aligned_data = GET_UNALIGNED_16(fldAddr);
-                    fldAddr                            = p_aligned_data;
-                }
-                break;
-
-            case TYP_INT:
-            case TYP_UINT:
-            case TYP_FLOAT:
-                static_assert_no_msg(sizeof(unsigned __int32) == sizeof(int));
-                static_assert_no_msg(sizeof(unsigned __int32) == sizeof(unsigned int));
-                static_assert_no_msg(sizeof(unsigned __int32) == sizeof(float));
-                if ((size_t)fldAddr % sizeof(unsigned __int32) != 0)
-                {
-                    *(unsigned __int32*)p_aligned_data = GET_UNALIGNED_32(fldAddr);
-                    fldAddr                            = p_aligned_data;
-                }
-                break;
-
-            case TYP_LONG:
-            case TYP_ULONG:
-            case TYP_DOUBLE:
-                static_assert_no_msg(sizeof(unsigned __int64) == sizeof(__int64));
-                static_assert_no_msg(sizeof(unsigned __int64) == sizeof(double));
-                if ((size_t)fldAddr % sizeof(unsigned __int64) != 0)
-                {
-                    *(unsigned __int64*)p_aligned_data = GET_UNALIGNED_64(fldAddr);
-                    fldAddr                            = p_aligned_data;
-                }
-                break;
-
-            default:
-                assert(!"Unexpected lclTyp");
-                break;
-        }
-    }
-#endif // DEBUG
-
-    switch (lclTyp)
+    GenTree* tree = nullptr;
+    switch (valueType)
     {
-        int     ival;
-        __int64 lval;
-        double  dval;
-
         case TYP_BOOL:
-            ival = *((bool*)fldAddr);
-            goto IVAL_COMMON;
-
         case TYP_BYTE:
-            ival = *((signed char*)fldAddr);
-            goto IVAL_COMMON;
-
         case TYP_UBYTE:
-            ival = *((unsigned char*)fldAddr);
-            goto IVAL_COMMON;
-
         case TYP_SHORT:
-            ival = *((short*)fldAddr);
-            goto IVAL_COMMON;
-
         case TYP_USHORT:
-            ival = *((unsigned short*)fldAddr);
-            goto IVAL_COMMON;
-
         case TYP_UINT:
         case TYP_INT:
-            ival = *((int*)fldAddr);
-        IVAL_COMMON:
-            op1 = gtNewIconNode(ival);
+            tree = gtNewIconNode((ssize_t)value);
             break;
 
         case TYP_LONG:
         case TYP_ULONG:
-            lval = *((__int64*)fldAddr);
-            op1  = gtNewLconNode(lval);
+            tree = gtNewLconNode((INT64)value);
             break;
 
         case TYP_FLOAT:
-            dval        = *((float*)fldAddr);
-            op1         = gtNewDconNode(dval);
-            op1->gtType = TYP_FLOAT;
+            tree = gtNewDconNode(*((float*)&value));
             break;
 
         case TYP_DOUBLE:
-            dval = *((double*)fldAddr);
-            op1  = gtNewDconNode(dval);
+            tree = gtNewDconNode(*((double*)&value));
```

Comment:
I don't think this is right on arm32 (and actually, it is UB on all platforms, but so was the existing code). `value` might not be 8-byte aligned while the double load needs 8-byte alignment. It was the reason for the old ceremony in this function.

The safest would be to go through `memcpy`.

Reply:
On second thought the alignment should be right given that `value` is an 8-byte integer. But it is still UB to reinterpret the value like this, so I think it would be better to go through `memcpy`.

Reply:
isn't it the same as was before?

Reply:
You probably mean the old ceremony for SPMI that actually did a memory load while we just convert a local, no?

Reply:
No, this was directly dereferencing the field address before. Now this is reinterpreting `ulong*` to a different kind of pointer. This is UB due to strict aliasing.

Reply:
> You probably mean the old ceremony for SPMI that actually did a memory load while we just convert a local, no?

Right, I mean the logic in this function that was ensuring alignment.

Reply:
I see, thanks